go-controller/pkg/ovn/logical_switch_cache.go

```diff
@@ -0,0 +1,168 @@
+// subnet allocations (v4 and v6) as well as the IPAM allocator instances for each
+// subnet managed for this node
+// logicalSwitchInfo is used to save switch specific information such as subnets,
+// allocated IPs per switch etc.
+type logicalSwitchInfo struct {
+	hostSubnets []*net.IPNet
+	ipams       []ipam.Interface
+}
+
+type logicalSwitchCache struct {
+	cache map[string]logicalSwitchInfo
+	// A RW mutex for logicalSwitchCache which holds logicalSwitch information
+	sync.RWMutex
+}
+
+type ipamFactoryFunc func(*net.IPNet) ipam.Interface
+
+func newIPAMAllocator(cidr *net.IPNet) ipam.Interface {
+	return ipam.NewCIDRRange(cidr)
+}
+
+func (cache *logicalSwitchCache) Add(nodeName string, hostSubnets []*net.IPNet, ipamFunc ipamFactoryFunc) error {
+	cache.Lock()
+	defer cache.Unlock()
+	if existing, ok := cache.cache[nodeName]; ok && !reflect.DeepEqual(existing.hostSubnets, hostSubnets) {
+		klog.Warningf("Node %q logical switch already in cache with subnet %s; replacing with %s", nodeName,
+			util.JoinIPNets(existing.hostSubnets, ","), util.JoinIPNets(hostSubnets, ","))
+	}
+	var ipams []ipam.Interface
+	for _, subnet := range hostSubnets {
+		ipam := ipamFunc(subnet)
+		if ipam == nil {
+			klog.Warningf("IPAM for subnet %s was not initialized for node %q", subnet, nodeName)
+			continue
+		}
+		// reserve IPs for gw, management port, hybrid overlay etc.
+		if err := reserveIps(subnet, ipam); err != nil {
+			klog.Warningf("Failed reserving IPs: %s", err)
+			continue
+		}
+		ipams = append(ipams, ipam)
+	}
+	cache.cache[nodeName] = logicalSwitchInfo{
+		hostSubnets: hostSubnets,
+		ipams:       ipams}
+
+	return nil
+}
+
+func (cache *logicalSwitchCache) Delete(nodeName string) error {
+	cache.Lock()
+	defer cache.Unlock()
+	delete(cache.cache, nodeName)
+	return nil
+}
+
+func (cache *logicalSwitchCache) IsNonHostSubnetSwitch(nodeName string) bool {
+	cache.RLock()
+	defer cache.RUnlock()
+	exists, ok := cache.cache[nodeName]
+	return ok && exists.hostSubnets == nil
+}
+
+func (cache *logicalSwitchCache) GetSwitchSubnet(nodeName string) *net.IPNet {
+	cache.RLock()
+	defer cache.RUnlock()
+	exists, ok := cache.cache[nodeName]
+	if ok && exists.hostSubnets != nil && len(exists.hostSubnets) > 0 {
+		//FIXME DUAL-STACK
+		subnet := *exists.hostSubnets[0]
+		return &subnet
+	}
+	return nil
+}
+
+func (cache *logicalSwitchCache) AllocateIP(nodeName string, ip *net.IP) error {
+	cache.RLock()
+	defer cache.RUnlock()
+	exists, ok := cache.cache[nodeName]
+	// FIXME DUAL-STACK
+	if ok && ip != nil && exists.ipams != nil && len(exists.ipams) > 0 {
+		if err := exists.ipams[0].Allocate(*ip); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (cache *logicalSwitchCache) AllocateNextIP(nodeName string) (*net.IPNet, error) {
+	cache.RLock()
+	defer cache.RUnlock()
+	exists, ok := cache.cache[nodeName]
+	// FIXME DUAL-STACK
+	if ok && exists.ipams != nil && len(exists.ipams) > 0 {
```

Comment:
Let's invert this one so that the final return is the success condition?

Reply:
But also, in what cases would ipams != nil && len(ipams) == 0? We should pick one; either nil or len(0) means no IPAM for this switch.

Reply:
paranoid programming. will change the check.

Reply:
Looks like ipams will be nil incase of not having an ipam

Reply:
Right, they would be, but I'm arguing that either: (a) `ipams` should always be a 0-length array or (b) `ipams` should either be nil, or always len >= 1

Reply:
Agree , I think that's the case here its nil or will have some valid ipam.